docs(rfc): deep links structure convention (#3878)
tari-project/tari: RFC/src/RFC-XXXX_DeepLinksConvencion.md

> # RFC-XXXX/DeepLinksConvencion

Reviewer: ```suggestion
# RFC-XXXX/DeepLinksConvention
```

Reviewer 2: Might I suggest making this RFC-1001

## Deep links structure convention.

![status: raw](theme/images/status-raw.svg)

**Maintainer(s)**: [Adrian Truszczyński](https://github.com/TruszczynskiA)

# Licence

[ The 3-Clause BSD Licence](https://opensource.org/licenses/BSD-3-Clause).

> Copyright 2202 The Tari Development Community

Reviewer: ```suggestion
Copyright 2022 The Tari Development Community
```

WHETHER IN CONTRACT, STRICT LIABILITY OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF
THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.

## Language

The keywords "MUST", "MUST NOT", "REQUIRED", "SHALL", "SHALL NOT", "SHOULD", "SHOULD NOT", "RECOMMENDED", 
"NOT RECOMMENDED", "MAY" and "OPTIONAL" in this document are to be interpreted as described in 
[BCP 14](https://tools.ietf.org/html/bcp14) (covering RFC2119 and RFC8174) when, and only when, they appear in all capitals, as 
shown here.

## Disclaimer

This document and its content are intended for information purposes only and may be subject to change or update
without notice.

This document may include preliminary concepts that may or may not be in the process of being developed by the Tari
community. The release of this document is intended solely for review and discussion by the community regarding the
technological merits of the potential system outlined herein.

## Goals

The aim of this Request for Comment (RFC) is to specify the deep links structure used in the Tari Aurora project.

> The primary motivation is to create a simple, human-readable, and scalable way to structure deep links used by the Tari Aurora clients.

Reviewer: Let's confirm but I suspect we can make this statement more general because I think this format will be used in future applications and not just Aurora

Author: Deep links should be unique and directed to the target app. e.g. on iOS, when one app will register to listen to the "tari" scheme then all other apps that using the same scheme will not be able to register to that callback till the first app will be removed.

## Description

Deep links are the URIs with hierarchical components sequence. We can use this sequence to pass and handle data in a standardized and predictable way. To do that, we need to pass three components to the target client:

> #### Scheme

Reviewer: ```suggestion
### Scheme
```

The scheme is used to address the client, which will handle the data from the command and data component. In the Tari Aurora project, we're using `tari` scheme to open the wallet app and handle provided data.

> #### Command

Reviewer: ```suggestion
### Command
```

> The command an optional component used to pass information about the action that should be performed by the client. The handler can use this info to determine which parser/decoder should be used to serialize data and what the app suppose to do with the data afterward.

Reviewer: ```suggestion
The command is an optional component used to pass information about the action that should be performed by the client. The handler can use this info to determine which parser/decoder should be used to deserialize data and what the app suppose to do with the data afterwards.
```

For simple actions, the command can be defined as a single phrase. For more complex actions, the command should be defined as a multi-path where the first path component should be the name of the action group.

**Examples**:

Simple Actions:
```
user_profile
login
```
Complex Actions:
```
payment/send
payment/request
```

> #### Data

Reviewer: ```suggestion
### Data
```

> The data component is a string of key-value pairs used by the parser/decoder to serialize the data, which the client will later use to handle the action. The data component should be formatted in the same way as a URL query. The sub-component should have a `?` prefix, the key-value pairs should be separated by `&`, and every key should be separated from the value by `=` character.

Reviewer: ```suggestion
The data component is a string of key-value pairs used by the parser/decoder to deserialize the data, which the client will later use to handle the action. The data component should be formatted in the same way as a URL query. The sub-component should have a `?` prefix, the key-value pairs should be separated by `&`, and every key should be separated from the value by `=` character.
```

> #### The Structure

Reviewer: ```suggestion
## The Structure
```

Author: I want to keep all sub-sections under the "Description" section which is H2. I will change the font size of sub-sections titles to H3.

Combining all three components, they will form a deep link with a structure presented below:
```
{scheme}://{command}?{data}
```
**Examples**:

> ```

Reviewer 2: Your code blocks need an `ignore` attribute so that mdbook test will pass 

https://rust-lang.github.io/mdBook/cli/test.html

shown here.

## Disclaimer

This document and its content are intended for information purposes only and may be subject to change or update
without notice.

This document may include preliminary concepts that may or may not be in the process of being developed by the Tari
community. The release of this document is intended solely for review and discussion by the community regarding the
technological merits of the potential system outlined herein.

## Goals

The aim of this Request for Comment (RFC) is to specify the deep links structure used in the Tari Aurora project.
The primary motivation is to create a simple, human-readable, and scalable way to structure deep links used by the Tari Aurora clients.

## Description

Deep links are the URIs with hierarchical components sequence. We can use this sequence to pass and handle data in a standardized and predictable way. To do that, we need to pass three components to the target client:

#### Scheme
The scheme is used to address the client, which will handle the data from the command and data component. In the Tari Aurora project, we're using `tari` scheme to open the wallet app and handle provided data.

#### Command

> The command an path string used to pass information about the action that should be performed by the client. The handler can use this info to determine which parser/decoder should be used to serialize data and what the app suppose to do with the data afterward.

Reviewer 3: ```suggestion
The command and path string is used to pass information about the action that should be performed by the client. The handler can use this info to determine which parser/decoder should be used to serialize data and what the app suppose to do with the data afterwards.
```

To support multiple domains/networks the first path component should contain a name of corresponded domain/network. The app before handle the data should first check that active account is pointed to the right domain/network and show the error is there a missmatch. 

For simple actions, the command can be defined as a single phrase. For more complex actions, the command should be defined as a multi-path where the second path component should be the name of the action group.

**Examples**:

Simple Actions:
```
mainnet/user_profile
testnet/login
```
Complex Actions:
```
mainnet/payment/send

> testnet/payment/request

Reviewer 4: Generally the resource is plural, which makes for (subjectively) nicer individual resource locators, e.g. /testnet/payments/<payment_id>

```suggestion
testnet/payments/request
```
